tough-kms/src/lib.rs

```diff
@@ -188,7 +196,130 @@ impl Sign for KmsRsaKey {
                 profile: self.profile.clone(),
                 key_id: self.key_id.clone(),
             })?;
-        let signature = response.signature.context(error::SignatureNotFound)?;
-        Ok(signature.to_vec())
+        let signature = response
+            .signature
+            .context(error::SignatureNotFound)?
+            .to_vec();
+
+        // sometimes KMS produces a signature that is shorter than the modulus. in those cases,
+        // we have observed that openssl and KMS will both validate the signature, but ring will
+        // not. if we pad the beginning of the signature with zeros to make the signature exactly
+        // the same length as the modulus, then ring will verify the signature.
+        let signature = pad_signature(signature, self.modulus_size_bytes)?;
+        Ok(signature)
+    }
+}
+
+/// Parses the `CustomerMasterKeySpec` string returned by KMS, e.g. `RSA_3072` and returns the size
+/// of the modulus in bytes. For example `RSA_3072` has a modulus of 3072 bits, so the function will
+/// return 384 == (3072 / 8). If the parsed number is not divisible by 8, an error is returned.
+fn parse_modulus_length_bytes(spec: &str) -> error::Result<usize> {
+    // only RSA is currently supported
+    ensure!(
+        spec.starts_with("RSA_"),
+        error::BadCustomerMasterKeySpec { spec }
+    );
```

Comment:
Instead of check for RSA prefix here do you think we can call this method only when signing algorithm is `KmsSigningAlgorithm.RsassaPssSha256` that way when we plan to support additional algorithm developer can handle different algorithm in `tuf_key` method.

Reply:
I could call it only when the algorithm is `KmsSigningAlgorithm.RsassaPssSha256` (looking into this now). But even so, I need to raise an error if the string isn't what I expect it to be. Even if we went with a variant as you suggest above, we'd be raising an error for unexpected strings here.

Reply:
It looks pretty ugly to me, especially since this is happening inside of a function that is a member of `KmsRsaKey`, so by definition it will only happen for RSA algorithms. Also the algorithm is a string.

```rust
        // sometimes KMS produces an RMS signature that is shorter than the modulus. in those cases,
        // we have observed that openssl and KMS will both validate the signature, but ring will
        // not. if we pad the beginning of the signature with zeros to make the signature exactly
        // the same length as the modulus, then ring will verify the signature.
        let signature = match self.signing_algorithm.as_str() {
            "I don't know exactly what this is right now" => {
                pad_signature(signature, self.modulus_size_bytes)?
            }
            _ => signature,
        };
```

Reply:
I feel it is a better check then checking for `RSA` prefix which is some kind of naming convention followed by KMS. I assume this check would be added for `parse_modulus_length_bytes` in tuf_key instead of `pad_signature`. Additionally it  helps us to throw error if user provided key type is not supported by us. For string comparison we can use `KmsSigningAlgorithm::RsassaPssSha256.value()`

Reply:
I think the `KmsRsaKey` should hold a `KmsSigningAlgorithm` instead of a string. Do you remember why you made it a string?

```rust
/// Implements the Sign trait for KMS rsa Key
pub struct KmsRsaKey {
    /// Key Id of Customer Managed Key in KMS used to sign the message
    key_id: String,
    /// Aws account profile
    profile: Option<String>,
    /// KmsClient Object to query AWS KMS
    client: Option<KmsClient>,
    /// Public Key corresponding to Customer Managed Key
    public_key: Decoded<RsaPem>,
    /// Signing Algorithm to be used for the Customer Managed Key
    signing_algorithm: String, // should be KmsSigningAlgorithm instead of String?
    /// The size of the RSA key modulus in bytes.
    modulus_size_bytes: usize,
}
```

Reply:
Seems like a tiny, non-breaking change to hold the variant instead of a string there, which then makes checking the algorithm before using the padding function much cleaner. I will push a version that does this.

Reply:
the `signing_algorithm` type is already `KmsSigningAlgorithm` https://github.com/awslabs/tough/blob/develop/tough-kms/src/lib.rs#L63. Not, sure how you saw it as a string.

Reply:
Take a look at [0124b97](https://github.com/awslabs/tough/compare/ffd3933d0f0b5c4ab1aeaa888eb0aa1c9374c2d5..0124b97dfc0db514c1d4fb463ca7315429933586)

Reply:
Ignore the above comment. You are correct.